iceoryx_utils/include/iceoryx_utils/internal/units/duration.hpp

```diff
@@ -186,49 +265,59 @@ class Duration
-    friend constexpr Duration duration_literals::operator"" _ns(long double);
     friend constexpr Duration duration_literals::operator"" _ns(unsigned long long int); // PRQA S 48
-    friend constexpr Duration duration_literals::operator"" _us(long double);
     friend constexpr Duration duration_literals::operator"" _us(unsigned long long int); // PRQA S 48
-    friend constexpr Duration duration_literals::operator"" _ms(long double);
     friend constexpr Duration duration_literals::operator"" _ms(unsigned long long int); // PRQA S 48
-    friend constexpr Duration duration_literals::operator"" _s(long double);
-    friend constexpr Duration duration_literals::operator"" _s(unsigned long long int); // PRQA S 48
-    friend constexpr Duration duration_literals::operator"" _m(long double);
-    friend constexpr Duration duration_literals::operator"" _m(unsigned long long int); // PRQA S 48
-    friend constexpr Duration duration_literals::operator"" _h(long double);
-    friend constexpr Duration duration_literals::operator"" _h(unsigned long long int); // PRQA S 48
-    friend constexpr Duration duration_literals::operator"" _d(long double);
-    friend constexpr Duration duration_literals::operator"" _d(unsigned long long int); // PRQA S 48
+    friend constexpr Duration duration_literals::operator"" _s(unsigned long long int);  // PRQA S 48
+    friend constexpr Duration duration_literals::operator"" _m(unsigned long long int);  // PRQA S 48
+    friend constexpr Duration duration_literals::operator"" _h(unsigned long long int);  // PRQA S 48
+    friend constexpr Duration duration_literals::operator"" _d(unsigned long long int);  // PRQA S 48
+
+    template <typename T>
+    friend constexpr Duration operator*(const T& left, const Duration& right);
+
+    friend std::ostream& operator<<(std::ostream& stream, const Duration& t);
 
   private:
-    /// @brief constructor needs to be private to ensure a unit safe usage of duration
-    constexpr explicit Duration(const long double durationInSeconds);
-    long double durationInSeconds{0.0};
+    template <typename T>
+    inline constexpr Duration
+    multiplySeconds(const uint64_t seconds, const std::enable_if_t<!std::is_floating_point<T>::value, T>& right) const;
+    template <typename T>
+    inline constexpr Duration multiplySeconds(const uint64_t seconds,
+                                              const std::enable_if_t<std::is_floating_point<T>::value, T>& right) const;
+    template <typename T>
+    inline constexpr Duration
+    multiplyNanoseconds(const uint32_t nanoseconds,
+                        const std::enable_if_t<!std::is_floating_point<T>::value, T>& right) const;
+    template <typename T>
+    inline constexpr Duration
+    multiplyNanoseconds(const uint32_t nanoseconds,
+                        const std::enable_if_t<std::is_floating_point<T>::value, T>& right) const;
+
+  private:
+    uint64_t m_seconds{0};
+    uint32_t m_nanoseconds{0};
```

Comment:
Next to the `double` implementation I found the inability to calculate _negative_ durations (like in received expected vs. actual).
In design you will have chosen for `unsigned` for reasons *)... But it came unexpected to me as there is no such limitation e.g. with `chrono`. (*No negative offsets for timers?)
Did I over-read the documentation explaining it (kindfully point me there, thanks~)?

Reply:
This was just a straight port to integer types without changing the previous behavior. @elfenpiff once told me why he did it this way, but I forgot it in the meantime. Maybe he can enlighten you.
Regarding the documentation, that's unfortunately missing. This is something we have to improve the next months.

Reply:
Duration is a measurement unit for the distance of two time points and it does not make sense to have a negative distance. It is like the distance of two points in space, negative distance is weird.

I think `std::chrono::duration` behaves identical when you call for instance `count`.

But maybe we have to rethink `duration`. For me `units::time` would make more sense somehow with a `units::duration` specialization. It would be very helpful for methods like this:
```cpp
mySemaphore.sleep_until(units::time pointInTIme);
mySemaphore.sleep_for(units::duration duration);
```

@elBoberido what do you think, would this be something for a follow up PR?

Reply:
Seems like Rust's also [not allowing](https://docs.rs/chrono/0.4.6/chrono/struct.Duration.html) a negative duration:

> As standard library implementation is limited to non-negative values.

I'm still unsure on why `std::chrono::duration` can be negative. We might ping @ [HowardHinnant](https://github.com/HowardHinnant) the creator of `chrono` :-)

Reply:
@elfenpiff I like the idea. The difference of two `time` objects would be a `duration`. It would need to be truncated to 0 for negative values. Maybe that's also the reason why `chrono::duration` supports negative values.

I think we should follow @mossmaurice suggestion and first fully explore if chrono is an option before we add more features.

@mossmaurice just for completeness. The link you added was to the `chrono` crate and that does allow negative durations. The standard library has `std::time::Duration` which does not allow negative durations, though.

Reply:
@elBoberido Ok, thanks for clarifying that. My Rust knowledge is still quite rusty 🤡

Reply:
@mossmaurice time to change that and make Ferris the 🦀 happy 😁

Reply:
Thanks all for giving inputs on this! Will follow the discussion interested~!